pandas/tests/reshape/test_concat.py

```diff
@@ -820,11 +821,59 @@ def test_append_preserve_index_name(self):
         result = df1.append(df2)
         assert result.index.name == 'A'
 
+    @pytest.mark.parametrize("df_columns", [
+        pd.RangeIndex(3),
+        pd.CategoricalIndex('A B C'.split()),
+        pd.MultiIndex.from_arrays(['A B C'.split(), 'D E F'.split()]),
+        pd.IntervalIndex.from_breaks([0, 1, 2, 3]),
+        pd.DatetimeIndex([dt.datetime(2013, 1, 3, 0, 0),
+                          dt.datetime(2013, 1, 3, 6, 10),
+                          dt.datetime(2013, 1, 3, 7, 12)]),
+        pd.Index([1, 2, 3]),
+    ])
+    def test_append_same_columns_type(self, df_columns):
+        # GH18359
+
+        # ser.index is a normal pd.Index, result from df.append(ser) should be
```

Comment:
make this as a separate test (the if part), though I suspect you are actually hiding a bug (see above).